exercises/diffie-hellman/src/test/java/DiffieHellmanTest.java

```diff
@@ -0,0 +1,95 @@
+import org.junit.Before;
+import org.junit.Ignore;
+import org.junit.Test;
+
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.assertEquals;
+
+import java.math.BigInteger;
+import java.util.*;
+
+public class DiffieHellmanTest {
+
+    private DiffieHellman diffieHellman;
+
+    @Before
+    public void setUp() {
+        diffieHellman = new DiffieHellman();
+    }
+
+    @Test
+    public void testPrivateKeyInRange() {
+        BigInteger primeP = BigInteger.valueOf(23);
+        List<BigInteger> privateKeys = new ArrayList<>();
+        for (int i = 0; i < 10; i++) privateKeys.add(diffieHellman.privateKey(primeP));
+
+        for (BigInteger privateKey : privateKeys) {
+            assertTrue((privateKey.compareTo(BigInteger.ONE) >= 0) && (privateKey.compareTo(primeP) < 0));
+        }
+    }
+
+    // Note: due to the nature of randomness, there is always a chance that this test fails
+    // Be sure to check the actual generated values
+    @Ignore("Remove to run test")
+    @Test
+    public void testPrivateKeyRandomlyGenerated() {
+        BigInteger primeP = BigInteger.valueOf(7919);
+        Set<BigInteger> privateKeys = new HashSet<>();
+        for (int i = 0; i < 5; i++) privateKeys.add(diffieHellman.privateKey(primeP));
+
+        assertEquals(5, privateKeys.size());
+    }
+
+    @Ignore("Remove to run test")
+    @Test
+    public void testPublicKeyCorrectlyCalculated() {
+        BigInteger primeP = BigInteger.valueOf(23);
+        BigInteger primeG = BigInteger.valueOf(5);
+        BigInteger privateKey = BigInteger.valueOf(6);
+        BigInteger expected = BigInteger.valueOf(8);
+
+        assertEquals(expected, diffieHellman.publicKey(primeP, primeG, privateKey));
```

Comment:
Does `assertEquals` work well with `BigInteger`? Does it give a good error message if the test fails?

Reply:
Yes it does. It prints expected value and actual value in numeric format.
![image](https://user-images.githubusercontent.com/21372947/36643177-97b7198c-1a48-11e8-9b9d-78981ec9f1be.png)